lib/factory_bot/definition.rb

```diff
@@ -114,14 +114,28 @@ def trait_by_name(name)
       trait_for(name) || Internal.trait_by_name(name)
     end
 
+    # This is a hot path when using traits so:
+    # - Memoize the defined_traits by name so we can lookup in a hash instead of array
+    # - Store both string/symbol names to avoid needing to cast on lookup
+    # - Use plain Hash since HashWithIndifferentAccess incurs significant overhead
     def trait_for(name)
-      defined_traits.detect { |trait| trait.name == name.to_s }
+      @defined_traits_by_name ||= defined_traits.each_with_object({}) do |trait, memo|
+        # This method was previously using `detect` which means if there were
+        # multiple traits the first was returned. It doesn't seem like you should be able
+        # to register duplicate traits but without this logic a test was breaking:
+        # spec/acceptance/traits_spec.rb:136 (the one with female_admin_judy)
+        next if memo[trait.name]
+        memo[trait.name] = trait
+        memo[trait.name.to_sym] = trait
+      end
+      @defined_traits_by_name[name]
```

Comment:
Any reason to not to_s both the storage of the memorized trait name and the passed in name?

Reply:
They mention "- Store both string/symbol names to avoid needing to cast on lookup" in the comments. I wouldn't assume we get a huge performance win from this, but we do save an allocation each time we look up a trait.

It does occur to me that this could be a breaking change if anyone is passing in objects that don't respond to `to_sym` as trait names. I don't imagine people are passing in much beyond strings and symbols, but it is still worth thinking about.

Reply:
Trait names are cast to a string on initialization so we should be ok on the latter concern: https://github.com/thoughtbot/factory_bot/blob/master/lib/factory_bot/trait.rb#L7

Reply:
Actually I might have a case where we would need to cast on lookup, going to try to repro with a test.

Reply:
OK so you can technically define a trait with any value and the name is cast `to_s` on creation so we need to also cast to a string on lookup. For example these are technically valid traits:
```ruby
trait 42 do
  some_field { nil }
end

trait [1,2,3] do
  some_field { nil }
end
```

I backfilled specs to exercise this behavior and updated the memoization/lookup to be simpler.